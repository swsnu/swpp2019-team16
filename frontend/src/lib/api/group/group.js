import client from '../client';

export function createGroup({ groupId, riders, driver, from, to }) {
  return client.post(`/api/v1/group`, { groupId, riders, driver, from, to });
}

export function acceptGroup({ groupId, driverId }) {
  return client.put(`/api/v1/group/${groupId}`, driverId);
}

export function onTaxi(riderId) {
  return client.put(`/api/v1/group/taxi/riders`, riderId);
}

export function departure({ groupId, departureTime }) {
  return client.put(`/api/v1/group/taxi/departure`, { groupId, departureTime });
}

export function fetchMyGroup(groupId) {
  return client.get(`/api/v1/group/${groupId}`);
}

export function notifyDriverLocation({ driverId, groupId, locationInfo }) {
  return client.put(`/api/v1/group/taxi/location`, {
    driverId,
    groupId,
    locationInfo,
  });
}

export function confirmCost({ groupId, totalCost }) {
  return client.put(`/api/v1/group/cost`, { groupId, totalCost });
}
