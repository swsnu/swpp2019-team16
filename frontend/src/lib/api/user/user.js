import client from '../client';

export function get({ id }) {
  return client.get(`/api/v1/user/${id}`);
}

export function updatePoint({ userId, point }) {
  return client.put(`/api/v1/user/${userId}`, { point });
}

export function riderOnTaxi({ riderId }) {
  return client.put(`/api/v1/user/taxi/rider/${riderId}`);
}

export function driverGoTaxi({ driverId }) {
  return client.put(`/api/v1/user/taxi/driver/${driverId}`);
}
