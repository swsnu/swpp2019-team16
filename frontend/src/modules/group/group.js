import { createAction, handleActions } from 'redux-actions';
import { takeLatest } from 'redux-saga/effects';
import * as groupAPI from '../../lib/api/group/group';
import createRequestSaga, {
  createRequestActionTypes,
} from '../../lib/createRequestSaga';

const [
  ACCEPT_GROUP,
  ACCEPT_GROUP_SUCCESS,
  ACCEPT_GROUP_FAILURE,
] = createRequestActionTypes('group/ACCEPT_GROUP');

const [ON_TAXI, ON_TAXI_SUCCESS, ON_TAXI_FAILURE] = createRequestActionTypes(
  'group/ON_TAXI',
);

const [
  DEPARTURE,
  DEPARTURE_SUCCESS,
  DEPARTURE_FAILURE,
] = createRequestActionTypes('group/DEPARTURE');

const [
  FETCH_MY_GROUP,
  FETCH_MY_GROUP_SUCCESS,
  FETCH_MY_GROUP_FAILURE,
] = createRequestActionTypes('group/FETCH_MY_GROUP');

const [
  NOTIFY_DRIVER_LOCATION,
  NOTIFY_DRIVER_LOCATION_SUCCESS,
  NOTIFY_DRIVER_LOCATION_FAILURE,
] = createRequestActionTypes('group/NOTIFY_DRIVER_LOCATION');

const [
  CONFIRM_COST,
  CONFIRM_COST_SUCCESS,
  CONFIRM_COST_FAILURE,
] = createRequestActionTypes('group/CONFIRM_COST');

export const GROUP_CREATED = 'group/GROUP_CREATED';
export const GROUP_UPDATED = 'group/GROUP_UPDATED';
export const UNLOAD_GROUP = 'group/UNLOAD_GROUP';
export const GROUP_COST_UPDATED = 'group/GROUP_COST_UPDATED';
export const INIT_COST_CONFIRMED = 'group/INIT_COST_CONFIRMED';

export const groupCreated = createAction(
  GROUP_CREATED,
  ({ groupId, from, to }) => ({
    groupId,
    from,
    to,
  }),
);

export const groupCostUpdated = createAction(
  GROUP_COST_UPDATED,
  ({ groupId, riderCost, cost }) => ({
    groupId,
    riderCost,
    cost,
  }),
);

export const groupUpdated = createAction(
  GROUP_UPDATED,
  ({ groupId, driver, riders, fromLocation, toLocation }) => ({
    groupId,
    driver,
    riders,
    from: fromLocation,
    to: toLocation,
  }),
);

export const unloadGroup = createAction(UNLOAD_GROUP);

export const acceptGroup = createAction(
  ACCEPT_GROUP,
  ({ groupId, driverId }) => ({
    groupId,
    driverId,
  }),
);
export const onTaxi = createAction(ON_TAXI, riderId => riderId);
export const departure = createAction(
  DEPARTURE,
  ({ groupId, departureTime }) => ({
    groupId,
    departureTime,
  }),
);
export const fetchMyGroup = createAction(FETCH_MY_GROUP, groupId => groupId);
export const notifyDriverLocation = createAction(
  NOTIFY_DRIVER_LOCATION,
  ({ driverId, groupId, locationInfo }) => ({
    driverId,
    groupId,
    locationInfo,
  }),
);
export const confirmCost = createAction(
  CONFIRM_COST,
  ({ groupId, totalCost }) => ({
    groupId,
    totalCost,
  }),
);
export const initCostConfirmed = createAction(INIT_COST_CONFIRMED);

export const acceptGroupSaga = createRequestSaga(
  ACCEPT_GROUP,
  groupAPI.acceptGroup,
);
export const onTaxiSaga = createRequestSaga(ON_TAXI, groupAPI.onTaxi);
export const departureSaga = createRequestSaga(DEPARTURE, groupAPI.departure);
export const fetchMyGroupSaga = createRequestSaga(
  FETCH_MY_GROUP,
  groupAPI.fetchMyGroup,
);
export const notifyDriverLocationSaga = createRequestSaga(
  NOTIFY_DRIVER_LOCATION,
  groupAPI.notifyDriverLocation,
);
export const confirmCostSaga = createRequestSaga(
  CONFIRM_COST,
  groupAPI.confirmCost,
);

export function* groupSaga() {
  yield takeLatest(ACCEPT_GROUP, acceptGroupSaga);
  yield takeLatest(ON_TAXI, onTaxiSaga);
  yield takeLatest(DEPARTURE, departureSaga);
  yield takeLatest(FETCH_MY_GROUP, fetchMyGroupSaga);
  yield takeLatest(NOTIFY_DRIVER_LOCATION, notifyDriverLocationSaga);
  yield takeLatest(CONFIRM_COST, confirmCostSaga);
}

const initialState = {
  group: null,
  error: null,
  costConfirmed: false,
};

const group = handleActions(
  {
    [GROUP_CREATED]: (state, { payload: group }) => ({
      ...state,
      group,
    }),

    [GROUP_UPDATED]: (state, { payload: group }) => ({
      ...state,
      group,
    }),

    [UNLOAD_GROUP]: state => ({
      ...state,
      group: null,
    }),

    [GROUP_COST_UPDATED]: (state, { payload: group }) => ({
      ...state,
      group: {
        ...state.group,
        ...group,
      },
    }),

    [ACCEPT_GROUP_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [ACCEPT_GROUP_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [ON_TAXI_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [ON_TAXI_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [DEPARTURE_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [DEPARTURE_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [FETCH_MY_GROUP_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),

    [FETCH_MY_GROUP_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [NOTIFY_DRIVER_LOCATION_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: group,
    }),
    [NOTIFY_DRIVER_LOCATION_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),

    [CONFIRM_COST_SUCCESS]: (state, { payload: group }) => ({
      ...state,
      group: {
        ...state.group,
        ...group,
      },
      error: null,
      costConfirmed: true,
    }),

    [CONFIRM_COST_FAILURE]: (state, { payload: error }) => ({
      ...state,
      error: error,
    }),
    [INIT_COST_CONFIRMED]: state => ({
      ...state,
      costConfirmed: initialState.costConfirmed,
    }),
  },
  initialState,
);

export default group;
