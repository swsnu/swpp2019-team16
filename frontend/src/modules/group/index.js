import group from './group';

export * from './group';
export default group;