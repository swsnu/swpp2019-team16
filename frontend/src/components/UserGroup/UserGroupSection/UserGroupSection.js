import React, { useState } from 'react';
import Timer from 'react-compound-timer';
import PropTypes from 'prop-types';
import { GroupPropTypes } from '../../../types/group';
import { makeStyles } from '@material-ui/styles';
import Map from '../../common/Map';
import MapPin from '../../common/MapPin';
import RiderList from '../RiderList';
import DriverInfo from '../DriverInfo';
import Button from '../../common/Button';
import { UserPropsTypes } from '../../../types/user';
import { Typography } from '@material-ui/core';

const INITIAL_TIME_LEFT = 10000;

const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'column',
  },
  map: {
    marginBottom: '2rem',
  },
  location: {
    marginTop: '2rem',
    marginLeft: '1rem',
    marginRight: '1rem',
    marginBottom: '2rem',

    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
  },
  locationTitle: {
    marginBottom: '0.7rem',
  },
  locationTo: {
    marginTop: '0.7rem',
    marginBottom: '0.7rem',
  },
  timer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: '1rem',
  },
  driver: {
    marginLeft: '1rem',
    marginRight: '1rem',
    marginBottom: '1rem',
  },
  riderList: {
    marginLeft: '1rem',
    marginRight: '1rem',
    marginBottom: '1rem',
  },
  footer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

UserGroupSection.propTypes = {
  user: UserPropsTypes.isRequired,
  group: GroupPropTypes.isRequired,
  onTaxiRidersList: PropTypes.arrayOf(
    PropTypes.shape({
      riderId: PropTypes.number,
    }),
  ),
  onClickOnTaxi: PropTypes.func.isRequired,
  onClickGoTaxi: PropTypes.func.isRequired,
};

// TODO: dynamic map pin
function UserGroupSection({
  user,
  group,
  onTaxiRidersList,
  onClickOnTaxi = () => {},
  onClickGoTaxi = () => {},
}) {
  const [onTaxied, setOnTaxied] = useState(false);
  const [timeover, setTimeover] = useState(false);
  const styles = useStyles();

  return (
    <div className={styles.root}>
      <div className={styles.map}>
        <Map
          zoom={15}
          center={{
            lat: 37.480126,
            lng: 126.952436,
          }}
          width={'100%'}
          height={'400px'}
        >
          <MapPin lat={37.480126} lng={126.952436} />
          <MapPin lat={37.477023} lng={126.961957} />
        </Map>
      </div>
      <div className={styles.location}>
        <div className={styles.locationTitle}>
          <Typography variant={'h4'}>We are going from</Typography>
        </div>
        <Typography variant={'h2'}>
          {group.from} ~ {group.to}
        </Typography>
      </div>
      <div className={styles.driver}>
        <DriverInfo driver={group.driver} />
      </div>
      <div className={styles.riderList}>
        <RiderList riders={group.riders} onTaxiRidersList={onTaxiRidersList} />
      </div>
      <div className={styles.timer}>
        <Timer
          initialTime={INITIAL_TIME_LEFT}
          direction="backward"
          checkpoints={[
            {
              time: 0,
              callback: () => {
                console.log('timeover!');
                setTimeover(true);
              },
            },
          ]}
        >
          {() => (
            <Typography variant={'h3'}>
              <Timer.Seconds /> seconds left, before taxi leave
            </Typography>
          )}
        </Timer>
      </div>
      <div className={styles.footer}>
        {user.user.user_type === 'rider' ? (
          <Button
            onClick={() => {
              setOnTaxied(true);
              onClickOnTaxi();
            }}
            disabled={onTaxied}
          >
            <Typography variant={'body1'}>On Taxi</Typography>
          </Button>
        ) : (
          <Button onClick={onClickGoTaxi} disabled={!timeover}>
            <Typography variant={'body1'}>Go</Typography>
          </Button>
        )}
      </div>
    </div>
  );
}

export default UserGroupSection;
