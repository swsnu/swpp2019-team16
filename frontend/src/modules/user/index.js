import user from './user';

export * from './user';
export default user;
