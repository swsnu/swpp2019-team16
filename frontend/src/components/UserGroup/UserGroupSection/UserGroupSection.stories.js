import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { storiesOf } from '@storybook/react';

import UserGroupSection from './UserGroupSection';
import { ThemeProvider } from '@material-ui/styles';
import theme from '../../../lib/styles/theme';
import { CssBaseline } from '@material-ui/core';
import { mockGroup } from '../../../types/__mock__/group';
import { mockRider } from '../../../types/__mock__/user';

storiesOf('UserGroup/UserGroupSection', module)
  .addDecorator(story => (
    <BrowserRouter>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {story()}
      </ThemeProvider>
    </BrowserRouter>
  ))
  .add('default', () => (
    <UserGroupSection
      user={mockRider}
      group={mockGroup}
      onCarpoolRequest={() => {}}
      googleMap={{}}
      driverInfo={{}}
    />
  ));
