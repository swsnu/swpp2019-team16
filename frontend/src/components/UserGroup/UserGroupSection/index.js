import UserGroupSection from './UserGroupSection';
export default UserGroupSection;
