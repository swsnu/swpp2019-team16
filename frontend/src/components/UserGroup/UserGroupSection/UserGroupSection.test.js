import React from 'react';
import { render } from '@testing-library/react';
import UserGroupSection from './UserGroupSection';
import { mockRider } from '../../../types/__mock__/user';
import { mockGroup } from '../../../types/__mock__/group';

describe('<UserGroup />', () => {
  it('SHOULD match with snapshot WHEN all data in', async () => {
    const { container } = render(
      <UserGroupSection
        user={mockRider}
        group={mockGroup}
        onTaxiRidersList={[]}
        onClickOnTaxi={() => {}}
        onClickGoTaxi={() => {}}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
