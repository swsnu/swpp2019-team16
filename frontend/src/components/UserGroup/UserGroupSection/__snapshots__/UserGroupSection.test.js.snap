// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserGroup /> SHOULD match with snapshot WHEN all data in 1`] = `
<div>
  <div
    class="sc-bwzfXH uTxCW"
  >
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2"
        title="My Group"
      >
        My Group
      </h2>
    </main>
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            This is where 
            GoogleMap
             will be
          </h5>
        </div>
      </div>
    </main>
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4"
      >
        Driver Info
      </h4>
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Name: 
        MockDriver
      </h5>
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Vehicle: 
        BMW
      </h5>
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Plate No.: 
      </h5>
    </main>
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4"
      >
        Group Info
      </h4>
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Riders: 
        [ rider1 ] 
        [ rider2 ] 
        [ rider3 ] 
        [ rider4 ] 
         
      </h5>
    </main>
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4"
      >
        Time left until departure: 3:00 
      </h4>
    </main>
    <main
      class="sc-bdVaJa ekJyAd"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-contained"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          On Taxi
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </main>
  </div>
</div>
`;
